fdbserver/storageserver.actor.cpp

```diff
@@ -3476,19 +3507,29 @@ ACTOR Future<Void> replaceInterface( StorageServer* self, StorageServerInterface
 	return Void();
 }
 
-ACTOR Future<Void> storageServer( IKeyValueStore* persistentData, StorageServerInterface ssi, Reference<AsyncVar<ServerDBInfo>> db, std::string folder, Promise<Void> recovered )
+ACTOR Future<Void> storageServer( IKeyValueStore* persistentData, StorageServerInterface ssi, Reference<AsyncVar<ServerDBInfo>> db, std::string folder, Promise<Void> recovered)
 {
 	state StorageServer self(persistentData, db, ssi);
 	self.folder = folder;
 	self.sk = serverKeysPrefixFor( self.thisServerID ).withPrefix(systemKeys.begin);  // FFFF/serverKeys/[this server]/
 	try {
 		state double start = now();
 		TraceEvent("StorageServerRebootStart", self.thisServerID);
-		bool ok = wait( self.storage.restoreDurableState() );
-		if (!ok) {
-			if(recovered.canBeSet()) recovered.send(Void());
-			return Void();
-		}
+
+		state Future<bool> dispose = memoryStoreRecover (persistentData, self.cx, self.thisServerID);
+        state Future<bool> restored = self.storage.restoreDurableState();
+        Void _ = wait(success(restored) || success(dispose));
+        if (restored.isReady()) {
+           dispose.cancel();
+            if(!restored.get()) {
+                // recovery finished before dispose, but return error
+                if (recovered.canBeSet()) recovered.send(Void());
+                return Void();
+            }
+        } else if (dispose.isReady() && dispose.get()){
+			TraceEvent("DisposeStorageServer", self.thisServerID);
+			throw worker_removed();
+        }
```

Comment:
You probably want to do 
```suggestion
        } else { UNSTOPPABLE_ASSERT(false); }
```

here. Or you could change the `memoryStoreRecover` actor return `Void` so you wouldn't need to check for the value.